gradle-plugin/src/main/java/com/squareup/hephaestus/plugin/HephaestusPlugin.kt

```diff
@@ -58,101 +78,129 @@ open class HephaestusPlugin : Plugin<Project> {
+            CheckMixedSourceSet(project, compileTask).disablePreciseJavaTrackingIfNeeded()
+
             compileTask.logger.info(
                 "Hephaestus: Use precise java tracking: ${compileTask.usePreciseJavaTracking}"
             )
           }
         }
   }
 
+  @Suppress("UnstableApiUsage")
   @ExperimentalStdlibApi
   private fun disableIncrementalKotlinCompilation(
-    project: Project
+    project: Project,
+    isAndroidProject: Boolean
   ) {
     project.tasks
-        .withType(KotlinCompile::class.java)
-        .all { compileTask ->
-          val isStubGeneratingTask = compileTask is KaptGenerateStubsTask
-
-          if (isStubGeneratingTask) {
-            check(
-                compileTask.name.startsWith("kaptGenerateStubs") &&
-                    compileTask.name.endsWith("Kotlin")
-            ) {
-              // It's better to verify this naming pattern and fail early when it changes.
-              // See also: https://github.com/JetBrains/kotlin/blob/897c48f97eac4e0a5190773cd7b5c31683d601eb/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/Kapt3KotlinGradleSubplugin.kt#L490
-              "It's expected that the stub generating task is named " +
-                  "kaptGenerateStubs{variant}{buildType}Kotlin."
-            }
-          }
-
-          val configurationNameForIncrementalChecks = when {
-            // Disable incremental compilation, if the compiler plugin dependency isn't up-to-date.
-            // This will trigger a full compilation of a module using Hephaestus even though its
-            // source files might not have changed. This workaround is necessary, otherwise
-            // incremental builds are broken. See https://youtrack.jetbrains.com/issue/KT-38570
-            !isStubGeneratingTask -> PLUGIN_CLASSPATH_CONFIGURATION_NAME
-
-            // For Java / Kotlin projects it doesn't change.
-            project.isKotlinJvmProject -> "compileClasspath"
-
-            // Disable incremental compilation for the stub generating task, if any of the
-            // dependencies in the compile classpath have changed. This will make sure that we pick
-            // up any change from a dependency when merging all the classes. Without this workaround
-            // we could make changes in any library, but these changes wouldn't be contributed to the
-            // Dagger graph, because incremental compilation tricked us.
-            else -> {
-              val variant = project.androidVariants()
-                  .findVariantForCompileTask(compileTask)
-
-              "${variant.name}CompileClasspath"
-            }
+        .withType(KaptGenerateStubsTask::class.java)
+        .configureEach { stubsTask ->
+          // Disable incremental compilation for the stub generating task. If any of the
+          // dependencies in the compile classpath have changed, then this we need to trigger the
```

Comment:
"Trigger the compiler plugin if any dependencies in the compile classpath have changed."

Reply:
Done.